lib/api/src/grpc/transport_channel_pool.rs

```diff
@@ -0,0 +1,68 @@
+use rand::seq::SliceRandom;
+use std::collections::HashMap;
+use std::time::Duration;
+use std::vec::Vec;
+use tonic::transport::{Channel, Error, Uri};
+
+/// Holds a pool of channels established for a set of URIs.
+/// Channel are shared by cloning them.
+/// Make the `pool_size` larger to increase throughput.
+#[derive(Default)]
+pub struct TransportChannelPool {
+    ip_to_pool: tokio::sync::RwLock<HashMap<Uri, Vec<Channel>>>,
+    pool_size: usize,
+    grpc_timeout: Duration,
+}
+
+impl TransportChannelPool {
+    pub fn new(p2p_grpc_timeout: Duration, pool_size: usize) -> Self {
+        Self {
+            ip_to_pool: Default::default(),
+            grpc_timeout: p2p_grpc_timeout,
+            pool_size,
+        }
+    }
+
+    pub async fn make_channel(grpc_timeout: Duration, uri: Uri) -> Result<Channel, Error> {
+        let endpoint = Channel::builder(uri)
+            .timeout(grpc_timeout)
+            .connect_timeout(grpc_timeout)
+            .keep_alive_while_idle(true);
+        // `connect` is using the `Reconnect` network service internally to handle dropped connections
+        endpoint.connect().await
+    }
+
+    /// Initialize a pool for the URI and return a clone of the first channel.
+    /// Does not fail if the pool already exist.
+    async fn init_pool_for_uri(&self, uri: Uri) -> Result<Channel, Error> {
+        let mut guard = self.ip_to_pool.write().await;
+        match guard.get(&uri) {
+            None => {
+                let mut channels = Vec::with_capacity(self.pool_size);
+                for _ in 0..self.pool_size {
+                    let channel = Self::make_channel(self.grpc_timeout, uri.clone()).await?;
+                    channels.push(channel);
+                }
+                let result = channels[0].clone();
+                guard.insert(uri.clone(), channels);
+                Ok(result)
+            }
+            Some(channels) => Ok(channels[0].clone()),
+        }
+    }
+
+    pub async fn get_pooled_channel(&self, uri: &Uri) -> Option<Channel> {
+        let guard = self.ip_to_pool.read().await;
+        guard
+            .get(uri)
+            .and_then(|channels| channels.choose(&mut rand::thread_rng()))
```

Comment:
A possible amelioration for the future, pick the channel with the least amount of request in-flight instead of a random draw.
I tried to do it already using the Tower's balancing feature but I got lost in trait bound hell.
https://docs.rs/tower/latest/tower/balance/index.html

Reply:
Should we crate an issue or put a `TODO`?

Reply:
I will create an issue!

Reply:
here it is https://github.com/qdrant/qdrant/issues/589